### ISSUE TITLE
[question] How exactly does the multiprocessing of PPO2 work?

### ISSUE THREAD

**Issue by tirafesi** [hill-a/stable-baselines#322]
How exactly does the multiprocessing of _PPO2_ work and how does it affect _serial timesteps_  and _total timesteps_?

In my case, `n_steps = 2048`. So every 2048 timesteps, PPO will update and the callback will be called, correct? Using 1 process, this happened roughly once every 6s.

But when I increased the number of processes, it got slower. However, the `fps` increased. And so did the `total_timesteps`. So if I used 2 processes instead of one, `total_timesteps` would increase twice as fast (so 4096 per update). Meanwhile, `serial_timesteps` would increase at the same rate (2048 per update) - which is now slower than usual, because the callback now takes 12s to be called.

I'm just super confused. I thought by increasing the amount of processes available the model would learn faster, but that doesn't seem to be the case? How exactly does the amount of processes affect the algorithm speed?

**Comment by araffin** [comment]
Hello,
that is a good question.
First of all, I think some metrics in PPO2 are badly reported (notably `total_timesteps` and `episode_reward` see #143 ), we would welcome a PR that solves those issues ;)

Then, to answer your question, I recommend you taking a look at:
https://github.com/hill-a/stable-baselines/blob/bddd1ab29d9b96f94087576dc81db39e710f446e/stable_baselines/ppo2/ppo2.py#L421-L422

and [VecEnv Documentation](https://stable-baselines.readthedocs.io/en/master/guide/vec_envs.html)

>So every 2048 timesteps, PPO will update and the callback will be called, correct? 

In fact, PPO will update every 2048 calls to the step function, so every 2048 x n steps when using n environments. You also need to account for that in choosing `nminibatches`.

Also, `serial_timesteps` is defined as:
https://github.com/hill-a/stable-baselines/blob/bddd1ab29d9b96f94087576dc81db39e710f446e/stable_baselines/ppo2/ppo2.py#L341

so it seems it is independent from the number of environments (not sure if it is a good thing...)

> thought by increasing the amount of processes available the model would learn faster, but that doesn't seem to be the case? How exactly does the amount of processes affect the algorithm speed?

To sum it up, with more processes, it goes faster (in term of FPS / wall clock time) and usually it improves exploration. However that usually don't make the algorithm more sample efficient.

**Comment by tirafesi** [comment]
Thank you! That was a pretty good answer, think I understand it now.

> and usually it improves exploration. However that usually don't make the algorithm more sample efficient.

Would this also be the case for PPO1 + MPI ?

**Comment by tirafesi** [comment]
Oh, another thing I don't think I quite understood:

> In fact, PPO will update every 2048 calls to the step function, so every 2048 x N steps when using N environments.

Since the environments are all running in parallel, shouldn't the update frequency be more or less the same, regardless of the of the amount of environments? Sure, if I have 2 envs then the update will be called after 4096 timesteps, but since these envs are running in parallel, then I can process timesteps in double the speed, so the overall frequency of the update calls should stay the same right?

This isn't happening though. If I have 2 envs, then the update will take twice as long to be called.

**Comment by araffin** [comment]
Environments are run in parallel (if you use `SubProcVecEnv`) but no asynchronously.
In addition to that, you have some communication costs too (that's what `DummyVecEnv` with n envs is sometimes faster than its subprocess counterpart), so doubling the number of processes does not scale linearly (see [Amdahl's law](https://en.wikipedia.org/wiki/Amdahl%27s_law)).

For those two reasons, augmenting the number of processes also increase the time between each update (even if the  number of FPS increases).

>Would this also be the case for PPO1 + MPI ?

what do you mean?

**Comment by tirafesi** [comment]
> that's why DummyVecEnv with n envs is sometimes faster than its subprocess counterpart

What's the point of using DummyVecEnv with multiple environments? Why would you want to have N envs if they're not running in parallel?

> What do you mean?

Using N envs in parallel, what would be more sample efficient (and thus have a faster speed of convergence)? PPO1 + MPI or PPO2? Why?

From your earlier answer, you said running envs in parallel doesn't actually increase sample efficiency, so does that mean a model training on a single env and a model training on N envs would both require more or less the same amount of time to reach the same performance level?

**Comment by H-Park** [comment]
> From your earlier answer, you said running envs in parallel doesn't actually increase sample efficiency, so does that mean a model training on a single env and a model training on N envs would both require more or less the same amount of time to reach the same performance level?

I assume you're asking about wall clock time: 

> To sum it up, with more processes, it goes faster (in term of FPS / wall clock time) and usually it improves exploration. However that usually don't make the algorithm more sample efficient.
